# all.position file is generated by plink
# SNPID RSID position
all.pos <- read.table('/statgen/ilaria/dyslexia/holland/06_info/all.positions',h=T)
# imputed dosage genotype values
# rows are samples(individuals) IDs, columns are SNPs 
snp <- readRDS('/statgen/ilaria/dyslexia/holland/00_data/holland_snp.common.order.RDS')
# genes reference hg19
# chromosome position_start position_end gene_name
genes <- read.table('/statgen/ilaria/dyslexia/refseq_hg19_all_forPlink.txt')

colnames(genes) <- c("CHR","start","end","GENE")
setwd('/statgen/ilaria/dyslexia/holland')

myrs.ids <- data.frame(RSID=(substr(colnames(snp),1, nchar(colnames(snp))-2)))

# In all.pos data.frame select only the SNPs present in your dataset (snp)
rs.pos <- merge(all.pos,myrs.ids,'RSID')

# generate a table with a snp in each row and the gene it is included in. If more genes are found for the same region
# they are attached with an underscore.
# RSID SNPIS position genes
rs.pos["genes"] <- NA
for (i in seq(1,nrow(rs.pos)))
{
  chr <- as.numeric(as.character(rs.pos[i,]$SNPID))
  pos <- as.numeric(as.character(rs.pos[i,]$position))
  gene <- as.character(genes[which(genes$CHR==chr & genes$start<=pos & genes$end>=pos),]$GENE)
  print(gene)
  if (length(gene)>0) 
  rs.pos[i,]$genes <- paste(gene, collapse = "_")
}

write.table(rs.pos,"./rs.pos.holland.txt",c=T,r=F,qu=F)
